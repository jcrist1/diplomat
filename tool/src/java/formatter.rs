use std::borrow::Cow;

use diplomat_core::{
    ast::Ident,
    hir::{
        self, FloatType, IntType, OutType, PrimitiveType, ReturnType, Slice, StructPathLike,
        SuccessType, TyPosition, Type, TypeContext, TypeId,
    },
};
use heck::ToLowerCamelCase;

use crate::c2::CFormatter;

pub(crate) struct JavaFormatter<'cx> {
    tcx: &'cx TypeContext,
    pub c: CFormatter<'cx>,
}

const INVALID_NAMES: &[&str] = &[
    "new", "static", "default", "private", "internal", "toString",
];

const DISALLOWED_CORE_TYPES: &[&str] = &["Object", "String"];

impl<'cx> JavaFormatter<'cx> {
    pub fn new(tcx: &'cx TypeContext) -> Self {
        Self {
            tcx,
            c: CFormatter::new(tcx),
        }
    }

    pub fn fmt_c_method_name<'a>(&self, ty: TypeId, method: &'a hir::Method) -> Cow<'a, str> {
        self.c.fmt_method_name(ty, method).into()
    }

    pub fn fmt_field_name<'a>(&self, field: &'a hir::StructField) -> Cow<'a, str> {
        let name = field.name.as_str().to_lower_camel_case();
        if INVALID_NAMES.contains(&&*name) {
            format!("{name}_").into()
        } else {
            name.into()
        }
    }
    pub fn fmt_method_name<'a>(&self, method: &'a hir::Method) -> Cow<'a, str> {
        let name = method.name.as_str().to_lower_camel_case();
        let name = method.attrs.rename.apply(name.into());
        if INVALID_NAMES.contains(&&*name) {
            format!("{name}_").into()
        } else {
            name
        }
    }

    pub fn fmt_return_type_java<'a>(&self, return_ty: &'a ReturnType) -> Cow<'a, str> {
        match return_ty {
            ReturnType::Infallible(ref success) => self.fmt_success_type_java(success),
            ReturnType::Fallible(ref success, _) => todo!(),
            ReturnType::Nullable(_) => todo!(),
        }
    }

    pub fn fmt_success_type_java<'a>(&self, success_ty: &'a SuccessType) -> Cow<'a, str> {
        match success_ty {
            SuccessType::Write => "String".into(),
            SuccessType::OutType(ref o) => self.fmt_java_type(o),
            SuccessType::Unit => "void".into(),
            _ => todo!(),
        }
    }

    pub fn fmt_native_type<'a, P: TyPosition>(&self, ty: &'a Type<P>) -> Cow<'a, str> {
        match ty {
            Type::Primitive(ref p) => self.fmt_primitive(p),
            Type::Opaque(_) => todo!(),
            Type::Struct(_) => todo!(),
            Type::Enum(_) => todo!(),
            Type::Slice(_) => todo!(),
            _ => todo!(),
        }
    }
    pub fn fmt_java_type<'a, P: TyPosition>(&self, ty: &'a Type<P>) -> Cow<'a, str> {
        match ty {
            hir::Type::Primitive(ref p) => self.fmt_primitive(p),
            hir::Type::Opaque(o) => self.tcx.resolve_opaque(o.tcx_id).name.to_string().into(),
            hir::Type::Struct(s) => self.tcx.resolve_type(s.id()).name().to_string().into(),
            hir::Type::Enum(e) => self.tcx.resolve_enum(e.tcx_id).name.to_string().into(),
            hir::Type::Slice(Slice::Str(_, _)) => "String".into(),
            hir::Type::Slice(Slice::Primitive(_, p)) => {
                format!("{}[]", self.fmt_primitive(p)).into()
            }
            hir::Type::Slice(Slice::Strs(_)) => "String []".into(),
            ty => todo!("haven't implemented, {ty:?}"),
        }
    }

    pub fn fmt_primitive<'a>(&self, ty: &'a PrimitiveType) -> Cow<'a, str> {
        match ty {
            PrimitiveType::Bool => "boolean",
            PrimitiveType::Char => "int",
            PrimitiveType::Byte => "byte",
            PrimitiveType::Int(IntType::I8 | IntType::U8) => "byte",
            PrimitiveType::Int(IntType::I16 | IntType::U16) => "short",
            PrimitiveType::Int(IntType::I32 | IntType::U32) => "int",
            PrimitiveType::Int(IntType::I64 | IntType::U64) => "long",
            PrimitiveType::IntSize(_) => "long",
            PrimitiveType::Int128(_) => panic!("128 not supported by java"),
            PrimitiveType::Float(FloatType::F32) => "float",
            PrimitiveType::Float(FloatType::F64) => "double",
        }
        .into()
    }

    pub fn fmt_param_name(&self, name: &str) -> String {
        name.to_lower_camel_case()
    }
}
