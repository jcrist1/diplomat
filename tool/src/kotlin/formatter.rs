use crate::c2::CFormatter;
use diplomat_core::hir::{self, TypeContext, TypeId};
use heck::ToLowerCamelCase;
use std::borrow::Cow;

/// This type mediates all formatting
///
/// All identifiers from the HIR should go through here before being formatted
/// into the output: This makes it easy to handle reserved words or add rename support
///
/// If you find yourself needing an identifier formatted in a context not yet available here, please add a new method
///
/// This type may be used by other backends attempting to figure out the names
/// of C types and methods.
pub(super) struct KotlinFormatter<'tcx> {
    c: CFormatter<'tcx>,
    strip_prefix: Option<String>,
}

const INVALID_METHOD_NAMES: &[&str] = &["new", "static", "default"];
const INVALID_FIELD_NAMES: &[&str] = &["new", "static", "default"];
const DISALLOWED_CORE_TYPES: &[&str] = &["Object", "String"];

impl<'tcx> KotlinFormatter<'tcx> {
    pub fn new(tcx: &'tcx TypeContext, strip_prefix: Option<String>) -> Self {
        Self {
            c: CFormatter::new(tcx),
            strip_prefix,
        }
    }

    pub fn fmt_docs(&self) -> String {
        // todo: fix this
        "Don't do it".to_string()
    }

    pub fn fmt_void(&self) -> &'static str {
        "Unit"
    }

    pub fn fmt_destructor_name(&self, id: TypeId) -> String {
        format!("{}_destroy", self.fmt_type_name(id))
    }

    pub fn fmt_string(&self) -> &'static str {
        "DiplomatStr"
    }

    pub fn fmt_c_method_name<'a>(&self, ty: TypeId, method: &'a hir::Method) -> Cow<'a, str> {
        self.c.fmt_method_name(ty, method).into()
    }

    /// Resolve and format a named type for use in diagnostics
    /// (don't apply rename rules and such)
    pub fn fmt_type_name_diagnostics(&self, id: TypeId) -> Cow<'tcx, str> {
        self.c.fmt_type_name_diagnostics(id)
    }

    pub fn fmt_primitive_as_ffi(&self, prim: hir::PrimitiveType) -> &'static str {
        use diplomat_core::hir::{FloatType, IntType, PrimitiveType};
        match prim {
            PrimitiveType::Bool => "Boolean",
            PrimitiveType::Char => "Char",
            PrimitiveType::Int(IntType::I8) => "Byte",
            PrimitiveType::Int(IntType::I16) => "Short",
            PrimitiveType::Int(IntType::I32) => "Int",
            PrimitiveType::Int(IntType::I64) => "Long",
            PrimitiveType::Int(IntType::U8) => "UByte",
            PrimitiveType::Int(IntType::U16) => "UShort",
            PrimitiveType::Int(IntType::U32) => "UInt",
            PrimitiveType::Int(IntType::U64) => "ULong",
            PrimitiveType::Byte => "Byte",
            PrimitiveType::IntSize(_) => "Long", // this feels wrong
            PrimitiveType::Float(FloatType::F32) => "Float",
            PrimitiveType::Float(FloatType::F64) => "Double",
            PrimitiveType::Int128(_) => panic!("i128 not supported in Dart"),
        }
    }

    pub fn fmt_method_name<'a>(&self, method: &'a hir::Method) -> Cow<'a, str> {
        // TODO(#60): handle other keywords

        // TODO: we should give attrs.rename() control over the camelcasing
        let name = method.name.as_str().to_lower_camel_case();
        let name = method.attrs.rename.apply(name.into());
        if INVALID_METHOD_NAMES.contains(&&*name) {
            format!("{name}_").into()
        } else {
            name
        }
    }

    pub fn fmt_setter_name(&self, method: &hir::Method) -> String {
        let name = &*self.fmt_method_name(method);
        let name = name.strip_prefix("set").unwrap().to_lower_camel_case();

        if INVALID_FIELD_NAMES.contains(&name.as_str()) {
            format!("{name}_")
        } else {
            name
        }
    }

    pub fn fmt_constructor_name(&self, method: &hir::Method) -> Option<String> {
        let mut name = self.fmt_method_name(method).into_owned();
        for prefix in ["try", "create", "new_", "new", "default_", "default", "get"] {
            name = name
                .strip_prefix(prefix)
                .map(|s| s.to_lower_camel_case())
                .unwrap_or(name);
        }

        if name.is_empty() {
            None
        } else if INVALID_METHOD_NAMES.contains(&name.as_str()) {
            Some(format!("{name}_"))
        } else {
            Some(name)
        }
    }

    pub fn fmt_param_name<'a>(&self, ident: &'a str) -> Cow<'a, str> {
        ident.to_lower_camel_case().into()
    }

    pub fn fmt_type_name(&self, id: TypeId) -> Cow<'tcx, str> {
        let resolved = self.c.tcx().resolve_type(id);

        let candidate: Cow<str> = if let Some(strip_prefix) = self.strip_prefix.as_ref() {
            resolved
                .name()
                .as_str()
                .strip_prefix(strip_prefix)
                .unwrap_or(resolved.name().as_str())
                .into()
        } else {
            resolved.name().as_str().into()
        };

        if DISALLOWED_CORE_TYPES.contains(&&*candidate) {
            panic!("{candidate:?} is not a valid Kotlin type name. Please rename.");
        }

        resolved.attrs().rename.apply(candidate)
    }

    pub fn fmt_pointer(&self) -> String {
        "Pointer".to_string()
    }

    pub fn fmt_usize(&self, cast: bool) -> &'static str {
        self.fmt_primitive_as_ffi(hir::PrimitiveType::IntSize(hir::IntSizeType::Usize))
    }

    pub fn fmt_nullable(&self, ident: &str) -> String {
        format!("{ident}?")
    }
}

#[cfg(test)]
pub mod test {
    use std::borrow::Cow;

    use super::KotlinFormatter;
    use diplomat_core::{
        ast::{self},
        hir::{self, TypeContext},
    };
    use proc_macro2::TokenStream;

    use quote::quote;

    pub fn new_tcx(tk_stream: TokenStream) -> TypeContext {
        let item = syn::parse2::<syn::File>(tk_stream).expect("failed to parse item ");

        let diplomat_file = ast::File::from(&item);

        let env = diplomat_file.all_types();
        let mut attr_validator = hir::BasicAttributeValidator::new("kotlin_test");
        attr_validator.support.renaming = true;
        attr_validator.support.disabling = true;

        match hir::TypeContext::from_ast(&env, attr_validator) {
            Ok(context) => context,
            Err(e) => {
                for err in e {
                    eprintln!("Lowering error: {}", err);
                }
                panic!()
            }
        }
    }

    #[test]
    fn test_type_name() {
        let tk_stream = quote! {
            #[diplomat::bridge]
            mod ffi {
                #[diplomat::opaque]
                struct MyOpaqueStruct {
                    a: SomeExternalType
                }

                #[diplomat::opaque]
                struct StringWrapper(String);

                impl MyOpaqueStruct {
                    pub fn new() -> Box<MyOpaqueStruct> {
                        unimplemented!();
                    }

                    pub fn get_byte(&self) -> u8 {
                        unimplemented!()
                    }

                    pub fn get_string_wrapper(&self) -> Box<StringWrapper> {
                        unimplemented!()
                    }

                }

            }
        };
        let tcx = new_tcx(tk_stream);
        let formatter = KotlinFormatter::new(&tcx, None);
        let opaques = tcx.opaques();
        assert!(!opaques.is_empty());
        let mut all_types = tcx.all_types();
        let (ty_id, _) = all_types.next().expect("Failed to get next type");

        assert_eq!(Cow::from("MyOpaqueStruct"), formatter.fmt_type_name(ty_id));

        let (ty_id, _) = all_types.next().expect("Failed to get next type");

        assert_eq!(Cow::from("StringWrapper"), formatter.fmt_type_name(ty_id));
    }
}
