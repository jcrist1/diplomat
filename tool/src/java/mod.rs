use std::borrow::Cow;
use std::collections::HashSet;
use std::fs::File;
use std::io::Write;
use std::path::Path;

use askama::Template;
use diplomat_core::hir::{
    self, OpaqueDef, ReturnType, Slice, SpecialMethod, StringEncoding, SuccessType, TypeContext,
    TypeDef, TypeId,
};
use formatter::JavaFormatter;

use crate::c2::{self};
use crate::common::{ErrorStore, FileMap};

const TMP_C_DIR: &str = "tmp";
const LIBRARY: &str = "somelib"; // todo: build from conf. Ensure that name is not the same as any
                                 // type
const GROUP: &str = "dev.diplomattest"; // todo: config
const _TMP_LIB_NAME: &str = "dev/diplomattest/somelib"; // todo: build from conf
const _JAVA_DIR: &str = "src/main/java/";

mod formatter;
pub fn run(
    tcx: &TypeContext,
    _conf_path: Option<&Path>,
    out_folder: &Path,
) -> std::io::Result<FileMap> {
    let files = FileMap::default();
    let mut context = c2::CContext::new(tcx, files, false);
    context.run();

    let errors = context.errors.take_all();

    if !errors.is_empty() {
        eprintln!("Found errors when generating c  code");
        for error in errors {
            eprintln!("\t{}: {}", error.0, error.1);
        }
    }

    let out_files = context.files.take_files();

    let tmp_path = out_folder.join(TMP_C_DIR);
    std::fs::create_dir(&tmp_path)?;
    let mut include_files = HashSet::new();
    for (subpath, text) in out_files {
        let out_path = tmp_path.join(&subpath);
        if !subpath.ends_with(".d.h") && subpath.ends_with(".h") {
            include_files.insert(subpath);
        }
        let parent = out_path
            .parent()
            .expect("Cannot create files at top level dir /");
        std::fs::create_dir_all(parent)?;
        let mut out_file = File::create(&out_path)?;
        out_file.write_all(text.as_bytes())?;
    }

    let lib_path = tmp_path.join(format!("{LIBRARY}.h"));

    let mut lib_file = File::create(&lib_path)?;
    for include in include_files {
        writeln!(lib_file, "#include \"{include}\"")?;
    }

    // jextract \
    //   --include-dir /path/to/mylib/include \
    //   --output src \
    //   --target-package org.jextract.mylib \
    //   --library mylib \
    //   /path/to/mylib/include/mylib.h

    let package = format!("{GROUP}.{LIBRARY}.ntv");
    let mut command = std::process::Command::new("jextract");
    command
        .arg("--include-dir")
        .arg(&tmp_path)
        .arg("--output")
        .arg(out_folder)
        .arg("--target-package")
        .arg(package)
        .arg("--library")
        .arg(LIBRARY)
        .arg(lib_path);

    println!("Running: {:?}", command);

    // todo: delete directory

    match command.output() {
        Err(err) => match err.kind() {
            std::io::ErrorKind::NotFound => {
                eprintln!("Check that jextract is in your path and all directories exist. See https://github.com/openjdk/jextract/blob/5715737be0a1a9de24cce3ee7190881cfc8b1350/doc/GUIDE.md");
                return Err(err);
            }
            _ => return Err(err),
        },
        Ok(ok) => {
            let stdout = String::from_utf8_lossy(&ok.stdout);
            println!("Output from jextract:\n{stdout}");

            let stderr = String::from_utf8_lossy(&ok.stderr);
            println!("Std Err from jextract:\n{stderr}");
        }
    }

    let java_formatter = JavaFormatter::new(tcx);
    let formatter = &java_formatter;
    let error_store = ErrorStore::default();
    let errors = &error_store;

    let ty_gen_cx = TyGenContext {
        tcx,
        formatter,
        errors,
    };

    let files = FileMap::default();
    for (id, ty) in tcx.all_types() {
        let _guard = ty_gen_cx.errors.set_context_ty(ty.name().as_str().into());
        if ty.attrs().disable {
            continue;
        }

        match ty {
            TypeDef::Opaque(o) => {
                // let type_name = o.name.to_string();

                // let (file_name, body) = ty_gen_cx.gen_opaque_def(o, id, GROUP, LIBRARY);

                // files.add_file(format!("src/main/kotlin/{file_name}"), body);
            }
            _ => continue,
        }
    }

    Ok(files)
}

#[derive(Clone, Debug)]
struct Param<'a> {
    name: Cow<'a, str>,
    ty: Cow<'a, str>,
}

#[derive(Clone, Debug)]
pub(crate) struct ParamConversion<'a> {
    converted_value: Cow<'a, str>,
    conversion_def: Cow<'a, str>,
}

#[derive(Template, Clone, Debug)]
#[template(path = "java/Method.java.jinja", escape = "none")]
pub(crate) struct MethodTpl<'a> {
    method_name: Option<Cow<'a, str>>,
    is_static: bool,
    return_ty: Cow<'a, str>,
    native_method: Cow<'a, str>,
    native_invocation: Cow<'a, str>,
    params: Vec<Param<'a>>,
    param_conversions: Vec<ParamConversion<'a>>,
    return_conversion: Cow<'a, str>,
    allocations: bool,
}

#[derive(Template, Clone, Debug)]
#[template(path = "java/Opaque.java.jinja", escape = "none")]
pub(crate) struct OpaqueTypeTpl<'a> {
    type_name: Cow<'a, str>,
    lib_name: Cow<'a, str>,
    domain: Cow<'a, str>,
    static_methods: Vec<Cow<'a, str>>,
    class_methods: Vec<Cow<'a, str>>,
}

#[derive(Template, Clone, Debug)]
#[template(path = "java/OpaqueReturn.java.jinja", escape = "none")]
pub(crate) struct OpaqueReturnTpl<'a> {
    return_ty: Cow<'a, str>,
}

#[derive(Template, Clone, Debug)]
#[template(path = "java/OpaqueConstructor.java.jinja", escape = "none")]
pub(crate) struct OpaqueConstructorTpl<'a> {
    return_ty: Cow<'a, str>,
}

struct TyGenContext<'a, 'cx> {
    tcx: &'cx TypeContext,
    formatter: &'a JavaFormatter<'cx>,
    errors: &'a ErrorStore<'cx, String>,
}

impl<'a, 'cx> TyGenContext<'a, 'cx> {
    fn gen_param_conversion<'b>(
        &self,
        param: &'b diplomat_core::hir::Param,
    ) -> ParamConversion<'b> {
        let diplomat_core::hir::Param { name, ty, .. } = param;
        // let java_ty = self.formatter.fmt_java_type(ty);
        let converted_value: Cow<'b, str> = format!("{name}Native").into();
        let (conversion, converted_value) = match ty {
            hir::Type::Primitive(_) => (name.as_str().into(), converted_value),
            hir::Type::Opaque(_) => (format!("{name}.inner").into(), converted_value),
            hir::Type::Struct(_) => todo!(),
            hir::Type::Enum(_) => todo!(),
            hir::Type::Slice(Slice::Str(_, StringEncoding::UnvalidatedUtf8)) => (
                format!(
                    r#"var {name}MemSeg = arena.allocateFrom({name}, StandardCharsets.UTF_8);
var {name}Len = {name}MemSeg.byteSize();"#
                )
                .into(),
                format!("{name}MemSeg, {name}Len").into(),
            ),
            hir::Type::Slice(Slice::Str(_, StringEncoding::UnvalidatedUtf16)) => (
                format!(
                    r#"var {name}MemSeg = arena.allocateFrom({name}, StandardCharsets.UTF_16);
var {name}Len = {name}MemSeg.byteSize();"#
                )
                .into(),
                format!("{name}MemSeg, {name}Len").into(),
            ),
            hir::Type::Slice(Slice::Str(_, StringEncoding::Utf8)) => (
                format!(
                    r#"var {name}MemSeg = arena.allocateFrom({name}, StandardCharsets.UTF_8);
var {name}Len = {name}MemSeg.byteSize();"#
                )
                .into(),
                format!("{name}MemSeg, {name}Len").into(),
            ),
            _ => todo!(),
        };
        ParamConversion {
            converted_value,
            conversion_def: conversion,
        }
    }

    fn gen_return_conversion(&self, ty: &ReturnType) -> Cow<'cx, str> {
        let ret = match ty {
            ReturnType::Infallible(ref ret) => ret,
            ReturnType::Fallible(ref ret, _) => ret,
            ReturnType::Nullable(ref ret) => ret,
        };
        let o = match ret {
            SuccessType::Write => return "".into(),
            SuccessType::OutType(ref o) => o,
            SuccessType::Unit => return "".into(),
            _ => todo!(),
        };
        match o {
            hir::Type::Primitive(o) => "return nativeVal;".into(),
            hir::Type::Opaque(o) => {
                let ty_name = &self.tcx.resolve_opaque(o.tcx_id).name;
                OpaqueReturnTpl {
                    return_ty: ty_name.as_str().into(),
                }
                .render()
                .unwrap_or_else(|err| {
                    panic!(
                        "Failed to render return val for type {}. Cause: {err}",
                        ty_name
                    )
                })
                .into()
            }
            hir::Type::Struct(_) => todo!(),
            hir::Type::Enum(_) => todo!(),
            hir::Type::Slice(_) => todo!(),
            _ => todo!(),
        }
    }

    fn gen_opaque_def(
        &self,
        o: &OpaqueDef,
        ty: TypeId,
        domain: &str,
        lib_name: &str,
    ) -> (Cow<str>, String) {
        let mut static_methods = Vec::new();
        let mut class_methods = Vec::new();
        for method in o.methods.iter() {
            let method_name = match method.attrs.special_method {
                // We need to reserve the default constructor for internal methods so a constructor
                // must always have params
                Some(SpecialMethod::Constructor) if !method.params.is_empty() => None,
                Some(SpecialMethod::Constructor) => {
                    eprintln!(
                        "Attempted to create constructor for {:?} type {:?}",
                        method.name, o.name
                    );
                    Some(self.formatter.fmt_method_name(method))
                }
                _ => Some(self.formatter.fmt_method_name(method)),
            };

            let return_ty = self.formatter.fmt_return_type_java(&method.output);
            let return_conversion = if method_name.is_none() {
                OpaqueConstructorTpl {
                    return_ty: return_ty.clone(),
                }
                .render()
                .unwrap_or_else(|err| {
                    panic!(
                        "Failed to render method {} for type {}. Cause: {err}",
                        method.name, o.name
                    )
                })
                .into()
            } else {
                self.gen_return_conversion(&method.output)
            };

            let allocations =
                method
                    .params
                    .iter()
                    .any(|diplomat_core::hir::Param { ty, .. }| {
                        matches!(ty, diplomat_core::hir::Type::Slice(_))
                    });
            let params = method
                .params
                .iter()
                .map(|diplomat_core::hir::Param { name, ty, .. }| Param {
                    name: self.formatter.fmt_param_name(name.as_str()).into(),
                    ty: self.formatter.fmt_java_type(ty),
                })
                .collect();
            let param_conversions: Vec<_> = method
                .param_self
                .iter()
                .map(|_| ParamConversion {
                    converted_value: "internal".into(),
                    conversion_def: "".into(),
                })
                .chain(
                    method
                        .params
                        .iter()
                        .map(|param| self.gen_param_conversion(param)),
                )
                .collect();
            let native_method: Cow<str> = format!(
                "{lib_name}_h.{}",
                self.formatter.fmt_c_method_name(ty, method)
            )
            .into();
            let native_invocation = if param_conversions.is_empty() {
                "nativeInvoker.apply".into()
            } else {
                native_method.clone()
            };
            let method_rendered = MethodTpl {
                method_name,
                is_static: method.param_self.is_none(),
                return_ty,
                native_method,
                native_invocation,
                params,
                param_conversions,
                return_conversion,
                allocations,
            }
            .render()
            .unwrap_or_else(|err| {
                panic!(
                    "Failed to render method {} for type {}. Cause: {err}",
                    method.name, o.name
                )
            });

            println!("HAve method!");
            println!("{method_rendered}");

            match method.param_self {
                Some(_) => class_methods.push(method_rendered.into()),
                None => static_methods.push(method_rendered.into()),
            }
        }

        let opaque_tpl = OpaqueTypeTpl {
            type_name: o.name.to_string().into(),
            lib_name: lib_name.into(),
            domain: domain.into(),
            static_methods,
            class_methods,
        };

        (
            format!("{}.java", o.name).into(),
            opaque_tpl.render().expect("Failed to render opaque type"),
        )
    }
}

#[cfg(test)]
mod test {

    use askama::Template;
    use diplomat_core::hir::TypeDef;
    use quote::quote;

    use crate::{common::ErrorStore, test::new_tcx};

    use super::{formatter::JavaFormatter, OpaqueTypeTpl, TyGenContext};
    #[test]
    fn test_opaque_render() {
        let opaque_type = OpaqueTypeTpl {
            type_name: "Opaque2".into(),
            lib_name: "somelib".into(),
            domain: "dev.diplomattest".into(),
            static_methods: Vec::new(),
            class_methods: Vec::new(),
        };

        let rendered = opaque_type.render().expect("Failed to render opaque type");
        insta::assert_snapshot!(rendered);
    }

    #[test]
    fn test_opaque() {
        let tk_stream = quote! {
            #[diplomat::bridge]
            mod ffi {

                #[diplomat::opaque]
                pub struct Opaque(String);

                impl Opaque {
                    #[diplomat::attr(supports = constructors, constructor)]
                    pub fn new() -> Box<Opaque> {
                        Box::new(Opaque("".into()))
                    }

                    pub fn from_str(input: &str) -> Box<Self> {
                        Box::new(Self(input.into()))
                    }


                    pub fn returns_usize() -> usize {
                        412
                    }

                    pub fn internal_len(&self) -> usize {
                        self.0.len()
                    }

                }
            }
        };

        let tcx = new_tcx(tk_stream);

        let formatter = JavaFormatter::new(&tcx);

        let errors = ErrorStore::default();
        let tcx_gen = TyGenContext {
            tcx: &tcx,
            formatter: &formatter,
            errors: &errors,
        };

        let (ty, some_opaque) = tcx
            .all_types()
            .next()
            .expect("Didn't find type despite there being one in quote");
        if let TypeDef::Opaque(opaque) = some_opaque {
            println!("Found opaque");
            let (_, rendered) = tcx_gen.gen_opaque_def(opaque, ty, "dev.diplomattest", "somelib");

            insta::assert_snapshot!(rendered);
        } else {
            panic!("Found type should have been opaque but wasn't");
        }
    }
}
