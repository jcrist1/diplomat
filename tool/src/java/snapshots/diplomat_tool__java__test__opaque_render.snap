---
source: tool/src/java/mod.rs
expression: rendered
---
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.*;


import java.lang.foreign.MemorySegment;
import java.lang.ref.Cleaner;


public class Opaque2 {

    MemorySegment internal;
    Cleaner.Cleanable cleanable;

    static class Opaque2Cleaner implements Runnable {

        MemorySegment segment;
        Opaque2Cleaner(MemorySegment segment) {
            this.segment = segment;
        }

        public void run() {
            somelib_h.Opaque2_destroy(this.segment);
        }
    }

    Opaque2() {}
    
    
}
